cmake-modules/CodeCoverage.cmake

```diff
@@ -0,0 +1,118 @@
+    endif()
+
+    ## Check if gcov and gcovr are available (gcovr generates html and xml from gcov files)
+    find_program(GCOV gcov)
+    find_program(GCOVR_PATH gcovr PATHS ${CMAKE_SOURCE_DIR}/scripts/test)
+    if(NOT GCOV)
+        message(FATAL_ERROR "gcov is required for code coverage reports.")
+    endif()
+    link_libraries(gcov)
+endif()
+
+# Set the compiler flags for the CMake project from where this is called (PARENT_SCOPE)
+function(append_code_coverage_for_current_project)
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -g -fprofile-arcs -ftest-coverage" PARENT_SCOPE)
+endfunction()
+
+
+# gcovr - html
+function(add_gcovr_html)
+
+    set(options NONE)
+    set(oneValueArgs TARGET_NAME EXECUTABLE_NAME)
+    set(multiValueArgs NA)
+    cmake_parse_arguments(args "${options}" "${oneValueArgs}" "${multiValueArgs}" ${ARGN})
+
+    set(BASEDIR ${PROJECT_SOURCE_DIR})
+
+    add_custom_target(${args_TARGET_NAME}
+        # Run tests
+        ${args_EXECUTABLE_NAME}
+
+        # Create folder
+        COMMAND ${CMAKE_COMMAND} -E make_directory ${PROJECT_BINARY_DIR}/${args_TARGET_NAME}
+
+        # Running gcovr
+        COMMAND ${GCOVR_PATH} --html --html-details
+            -r ${BASEDIR}
+            --object-directory=${PROJECT_BINARY_DIR}
+            -o ${args_TARGET_NAME}/index.html
+
+        BYPRODUCTS ${PROJECT_BINARY_DIR}/${args_TARGET_NAME}  # report directory
+        WORKING_DIRECTORY ${PROJECT_BINARY_DIR}
+        DEPENDS
+        VERBATIM
+        COMMENT "Running gcovr to produce HTML code coverage report."
+    )
+
+    # Show info where to find the report
+    add_custom_command(TARGET ${args_TARGET_NAME} POST_BUILD
+        COMMAND ;
+        COMMENT "Open ./${args_TARGET_NAME}/index.html in your browser to view the coverage report."
+    )
+
+endfunction()
+
+# gcovr - xml
+function(add_gcovr_xml)
+
+    set(options NONE)
+    set(oneValueArgs TARGET_NAME EXECUTABLE_NAME)
+    set(multiValueArgs NA)
+    cmake_parse_arguments(args "${options}" "${oneValueArgs}" "${multiValueArgs}" ${ARGN})
+
+    set(BASEDIR ${PROJECT_SOURCE_DIR})
+
+    if (NOT DEFINED ENV{AZURE_CI_TEST})
+        set(RUN_EXE ${args_EXECUTABLE_NAME})
+    endif()
+    
+    add_custom_target(${args_TARGET_NAME}
+        # Running on CI won't require to run tests exe since it was run on previous step
+        ${RUN_EXE}
+
+        # Running gcovr
+        COMMAND ${GCOVR_PATH} --xml
+            -r ${BASEDIR}
+            --object-directory=${PROJECT_BINARY_DIR}
+            -o ${args_TARGET_NAME}.xml
+        BYPRODUCTS ${args_TARGET_NAME}.xml
+        WORKING_DIRECTORY ${PROJECT_BINARY_DIR}
+        DEPENDS
+        VERBATIM
+        COMMENT "Running gcovr to produce Cobertura code coverage report."
+    )
+
+    # Show info where to find the report
+    add_custom_command(TARGET ${args_TARGET_NAME} POST_BUILD
+        COMMAND ;
+        COMMENT "Cobertura code coverage report saved in ${args_TARGET_NAME}.xml."
```

Comment:
What is Cobertura?

Reply:
that's the formatting-type of the code coverage XML report we generate.
`Cobertura` is mainly used for Java. It defines how the XML is organized. We use this formatting for the XML so we can export it to Azure DevOps.
There are other formats that gcovr (the tool we use) can generate, but Azure DevOps might not.